homeassistant/components/sensor/recollect_waste.py

```diff
@@ -0,0 +1,100 @@
+"""
+Support for Recollect Waste curbside collection pickup.
+
+For more details about this platform, please refer to the documentation at
+https://www.home-assistant.io/components/sensor.recollect_waste/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (CONF_NAME, CONF_SCAN_INTERVAL)
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+REQUIREMENTS = ['recollect-waste==1.0.1']
+
+_LOGGER = logging.getLogger(__name__)
+ATTR_PICKUP_TYPES = 'pickup_types'
+ATTR_AREA_NAME = 'area_name'
+CONF_PLACE_ID = 'place_id'
+CONF_SERVICE_ID = 'service_id'
+DEFAULT_SCAN_INTERVAL = timedelta(days=1)
+DOMAIN = 'recollect_waste'
+ICON = 'mdi:trash-can-outline'
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_PLACE_ID): cv.string,
+    vol.Required(CONF_SERVICE_ID): cv.string,
+    vol.Optional(CONF_NAME, default=DOMAIN): cv.string,
+    vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): (
+        vol.All(cv.time_period, cv.positive_timedelta))
+})
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Recollect Waste platform."""
+    add_entities([RecollectWasteSensor(
+        config.get(CONF_NAME),
+        config[CONF_PLACE_ID],
+        config[CONF_SERVICE_ID],
+        config.get(CONF_SCAN_INTERVAL))], True)
+
+
+class RecollectWasteSensor(Entity):
+    """Recollect Waste Sensor."""
+
+    def __init__(self, name, place_id, service_id, scan_interval):
+        """Initialize the sensor."""
+        self._attributes = {}
+        self._name = name
+        self._state = None
+        self.place_id = place_id
+        self.service_id = service_id
+        self.update = Throttle(scan_interval)(self._update)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return "{}{}".format(self.place_id, self.service_id)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return ICON
+
+    def _update(self):
+        """Update device state."""
+        import recollect_waste
+
+        try:
+            # pylint: disable=no-member
+            client = recollect_waste.RecollectWasteClient(self.place_id,
```

Comment:
Why do we need to create a new client each update? Can't we create the client early and save it as an instance attribute on the entity?

Reply:
If the config options can cause an exception, we should check this already in setup platform and fail the platform setup on error.

Reply:
The application won't know if the config options will create an exception until the first time the client is used to retrieve the data. Should I use the client during setup to test it then and do nothing with the results? I could set the values retrieved if successful on the Enity. Seems like unnecessary duplicate logic but it would mean the error would occur during setup instead of waiting for the default 1 day before knowing the config was wrong.

Reply:
Actually the update occurs  when Home Assistant is restarted so what I said above is not an issue. Also Home Assistant checks required configuration properties and handles those cases?

Reply:
Yes, we want to validate the client during setup. Please make that change. If wanted, we can pass a cache of the result to the entity when instantiating it.

Reply:
Home assistant can't check that the client has gotten the correct arguments, since we don't validate those specifically, only generally. We also can't check beforehand, unless we try, that it can reach the api etc. We should always check this during setup.

Reply:
Sounds good. I will make the change.